src/Microsoft.AspNet.Mvc.Localization/HtmlLocalizer.cs

```diff
@@ -0,0 +1,132 @@
+using Microsoft.Framework.WebEncoders;
+
+namespace Microsoft.AspNet.Mvc.Localization
+{
+    /// <summary>
+    /// An <see cref="IHtmlLocalizer"/> that uses the <see cref="IStringLocalizer"/> to provide localized HTML content.
+    /// </summary>
+    public class HtmlLocalizer : IHtmlLocalizer
+    {
+        private IStringLocalizer _localizer;
+        private readonly IStringLocalizerFactory _localizerFactory;
+        private readonly IHtmlEncoder _encoder;
+
+        /// <summary>
+        /// Creates a new <see cref="HtmlLocalizer"/>.
+        /// </summary>
+        /// <param name="factory">The <see cref="IStringLocalizerFactory"/>.</param>
+        /// <param name="encoder">The <see cref="IHtmlEncoder"/>.</param>
+        public HtmlLocalizer([NotNull] IStringLocalizerFactory factory, [NotNull] IHtmlEncoder encoder)
+        {
+            _localizerFactory = factory;
+            _encoder = encoder;
+        }
+
+        /// <summary>
+        /// Creates a new <see cref="HtmlLocalizer"/>.
+        /// </summary>
+        /// <param name="localizer">The <see cref="IStringLocalizer"/> to read strings from.</param>
+        /// <param name="encoder">The <see cref="IHtmlEncoder"/>.</param>
+        public HtmlLocalizer([NotNull] IStringLocalizer localizer, [NotNull] IHtmlEncoder encoder)
+        {
+            _localizer = localizer;
+            _encoder = encoder;
+        }
+
+        /// <inheritdoc />
+        public virtual LocalizedString this[[NotNull] string key] => _localizer[key];
+
+        /// <inheritdoc />
+        public virtual LocalizedString this[[NotNull] string key, params object[] arguments] =>
+            _localizer[key, arguments];
+
+        /// <summary>
+        /// Creates a new <see cref="IStringLocalizer"/> for a specific <see cref="CultureInfo"/>.
+        /// </summary>
+        /// <param name="culture">The <see cref="CultureInfo"/> to use.</param>
+        /// <returns>A culture-specific <see cref="IStringLocalizer"/>.</returns>
+        IStringLocalizer IStringLocalizer.WithCulture([NotNull] CultureInfo culture) => WithCulture(culture);
+
+        /// <inheritdoc />
+        public virtual LocalizedString GetString([NotNull] string key) => _localizer.GetString(key);
+
+        /// <inheritdoc />
+        public virtual LocalizedString GetString([NotNull] string key, params object[] arguments) =>
+            _localizer.GetString(key, arguments);
+
+        /// <inheritdoc />
+        public IEnumerable<LocalizedString> GetAllStrings(bool includeAncestorCultures) =>
+            _localizer.GetAllStrings(includeAncestorCultures);
+
+        /// <inheritdoc />
+        public virtual LocalizedHtmlString Html([NotNull] string key) => ToHtmlString(_localizer.GetString(key));
+
+        /// <inheritdoc />
+        public virtual LocalizedHtmlString Html([NotNull] string key, params object[] arguments) =>
+            ToHtmlString(_localizer.GetString(key, EncodeArguments(arguments)));
```

Comment:
As I said in a previous iteration, it is unsafe to assume the localized strings are HTML encoded. It's a bit ugly here because `_localizer.GetString()` does formatting and arguments that are already encoded shouldn't be encoded again. Probably need to do the formatting here instead because the underlying `IStringLocalizer` shouldn't be aware of HTML encoding.

The bottom line is: Do not instantiate an `HtmlString` containing anything other than encoded `string`s and explicitly-marked values from other `HtmlString`s. (Resources are not a trusted source of HTML `string`s because they may be used for multiple purposes.) This is a must-fix issue.